### PR TITLE
Refactor AgentSigner to make it more testable

### PR DESCRIPTION
**Pull request by davidxia**
Add loads more tests.

### CODE REVIEW

**Comment by davidxia** [comment]
@gimaker @rohansingh @mattnworb

**Comment by davidxia** [comment]
@nresare

**Comment by nresare** [comment]
I don't have time to review this at the moment, unfortunately :(

File: src/main/java/com/spotify/crtauth/agentsigner/AgentSigner.java [hunk 1]
```diff
@@ -71,21 +61,33 @@ public AgentSigner() throws CrtAuthException {
             "The environment variable SSH_AUTH_SOCK is not set. Please configure your ssh-agent.");
       }
 
-      channel = UnixSocketChannel.open(new UnixSocketAddress(new File(socketPath)));
+      // TODO (dxia) Do I need to keep a reference to this to close later?
```
Comment (mattnworb) [inline src/main/java/com/spotify/crtauth/agentsigner/AgentSigner.java]: looks like the inputstream/outputstream returned by `Channels.newOutputStream(..)` will close the input channel when you call `stream.close()`, which you close (via the AgentInput/OutputStream) in this class's `close` method.

File: src/main/java/com/spotify/crtauth/agentsigner/AgentSigner.java [hunk 2]
```diff
@@ -100,37 +102,14 @@ public AgentSigner() throws CrtAuthException {
...
-      final byte[] data2 = new byte[bytesToRead];
-      if (inputStream.read(data2, 0, bytesToRead) == -1) {
-        throw new CrtAuthException(
-            "Error reading SSH2_AGENTC_SIGN_REQUEST response from ssh-agent.");
-      }
-
-      final Iterator<byte[]> fields = new ByteIterator(data2);
-      final byte[] responseType = fields.next();
+      final byte[] responseType = iterator.next();
       assert new String(responseType).equals(RSA.RSA_LABEL);
```
Comment (mattnworb) [inline src/main/java/com/spotify/crtauth/agentsigner/AgentSigner.java]: I think this should be promoted to an actual if-else / throw , it's unlikely the user of this code will have assertions enabled

File: src/main/java/com/spotify/crtauth/agentsigner/IdentitiesAnswer.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+  private final int count;
+
+  private IdentitiesAnswer(final int length, final int responseCode, final int count) {
+    this.length = length;
+    this.responseCode = responseCode;
+    this.count = count;
+  }
+
+  static IdentitiesAnswer from(final byte[] bytes) {
+    // TODO (dxia) Should I use assert here or something else?
```
Comment (mattnworb) [inline src/main/java/com/spotify/crtauth/agentsigner/IdentitiesAnswer.java]: I vote for something else :smile:

File: src/main/java/com/spotify/crtauth/agentsigner/SignResponse.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    // TODO (dxia) Should I use assert here or something else?
+    assert bytes.length == 9;
+
+    // First four bytes represents length in bytes of rest of message
+    final int length = first(bytes);
+    // Next byte is the response code
+    final int responseCode = second(bytes);
+    assert responseCode == SSH2_AGENT_SIGN_RESPONSE;
+    // Next four bytes is the response length
+    final int responseLength = third(bytes);
```
Comment (mattnworb) [inline src/main/java/com/spotify/crtauth/agentsigner/SignResponse.java]: this code snippet is same as in `IdentitiesAnswer`, can the two be merged somehow?

**Comment by mattnworb** [comment]
I like the extraction out of byte-reading/protocol-handling from AgentSinger into the two new classes

I'd suggest renaming those classes though as something suffixed with InputStream / OutputStream that does not actually extend that interface might be confusing.  How about just `AgentInput` and `AgentOutput`?

:+1: if you replace the `assert`s with actual throws

**Comment by davidxia** [comment]
@mattnworb Thanks. That was very helpful.

File: src/main/java/com/spotify/crtauth/agentsigner/SignResponse.java (hunk 4 above)
Comment (davidxia) [inline src/main/java/com/spotify/crtauth/agentsigner/SignResponse.java]: They are almost the same except for checking the response code and the error messages. I'll leave it for now and try to come up with something nicer later.

**Pull request merged**